Description: Monkey-patch ipywidgets in the way that Sage upstream expects
 Sage upstream patches ipywidgets, but this is still being discussed
 upstream (see Bug below). In the meantime, we can have Sage monkey-patch
 ipywidgets at runtime, to prevent other users of ipywidgets being affected.
Author: Ximin Luo <infinity0@debian.org>
Author: Jeroen Demeyer <jdemeyer@cage.ugent.be>
Bug: https://github.com/jupyter-widgets/ipywidgets/pull/1031
---
This patch header follows DEP-3: http://dep.debian.net/deps/dep3/
Index: sagemath/src/sage/repl/ipython_kernel/widgets.py
===================================================================
--- sagemath.orig/src/sage/repl/ipython_kernel/widgets.py
+++ sagemath/src/sage/repl/ipython_kernel/widgets.py
@@ -26,6 +26,7 @@ We need to setup a proper test environme
 from ipywidgets.widgets import (IntSlider, IntRangeSlider,
                                 FloatSlider, FloatRangeSlider, Text,
                                 Textarea, ColorPicker, HTMLMath, Label,
+                                Widget,
                                 HBox, VBox, ValueWidget)
 from traitlets import List, Unicode, link
 
@@ -34,6 +35,36 @@ from sage.repl.user_globals import get_g
 from sage.misc.lazy_import import lazy_import
 lazy_import("sage.plot.colors", "Color")
 
+def __widgets__repr__(self):
+    """
+    Textual representation of this widget, mainly used for testing
+    and debugging.
+    """
+    # List of attributes to show with default values. The attribute
+    # is printed if it exists and its value is not equal to this
+    # default value.
+    attributes = [("value", None),
+                  ("min", None),
+                  ("max", None),
+                  ("step", None),
+                  ("options", None),
+                  ("description", ""),
+                  ("children", [])]
+    r = ""
+    for (attr, default) in attributes:
+        try:
+            v = getattr(self, attr)
+        except AttributeError:
+            continue
+        if default == v:
+            continue
+        if r:
+            r += ", "
+        r += attr + "=" + repr(v)
+    return "%s(%s)" % (type(self).__name__, r)
+Widget.__repr__ = __widgets__repr__
+del __widgets__repr__
+
 
 class HTMLText(HTMLMath):
     """
